xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/internal/doc/XWikiAttachmentList.java

```diff
@@ -0,0 +1,250 @@
+        if (removedAttachment != null) {
+            removedAttachment = super.remove(index);
+            document.setMetaDataDirty(true);
+        }
+        return removedAttachment;
+    }
+
+    /**
+     * Removes XWikiAttachment.
+     * 
+     * @param attachment XWikiAttachment to remove.
+     * @return true unless the attachment is not found
+     * @since 10.0RC1
+     */
+    @Override
+    public boolean remove(Object attachment)
+    {
+        XWikiAttachment xwikiAttachment = (XWikiAttachment) attachment;
+        if(map.put(xwikiAttachment.getFilename(), xwikiAttachment) == null)
+            return false;
+        map.remove(attachment);
+        updateMap();
+        return true;
+    }
+
+    /**
+     * Adds or replaces attachment with the same filename as the parameter.
+     * 
+     * @param attachment the attachment to add to the list
+     * @return the attachment that was added to the list in order of filename
+     * @since 10.0RC1
+     */
+    public XWikiAttachment set(XWikiAttachment attachment)
+    {
+        map.put(attachment.getFilename(), attachment);
+        added(attachment);
+        updateMap();
+        return attachment;
+    }
+
+    /**
+     * Adds or replaces attachment with the same filename as the parameter.
+     * 
+     * @param index this parameter is not used but is needed to override the method
+     * @param attachment the attachment to add to the list
+     * @return the attachment that was added to the list in order of filename
+     * @since 10.0RC1
+     */
+    @Override
+    public XWikiAttachment set(int index, XWikiAttachment attachment)
+    {
+        return set(attachment);
+    }
+
+    /**
+     * @param filename the filename of the attachment to be returned.
+     * @return attachment with the given filename or null if not found.
+     * @since 10.0RC1
+     */
+    public XWikiAttachment getByFilename(String filename)
+    {
+        return map.get(filename);
+    }
+
+    @Override
+    public boolean removeAll(Collection<?> c)
+    {
+        boolean changed = false;
+        for (XWikiAttachment x : (Collection<? extends XWikiAttachment>) c) {
+            if (this.list.contains(x)) {
+                if(remove(x)) {
+                    changed = true;
+                }
+            }
+        }
+        return changed;
+    }
+
+    @Override
+    public boolean retainAll(Collection<?> c)
+    {
+        boolean changed = false;
+        for (XWikiAttachment x : this.list) {
+            if (!((Collection<? extends XWikiAttachment>) (c)).contains(x)) {
+                if(remove(x)) {
+                    changed = true;
+                }
+            }
+        }
+        return changed;
+    }
+
+    /** Called when the list is updated. The method will be called at least once, but may be called several times */
+    public void onUpdate()
+    {
+        document.setMetaDataDirty(true);
+    }
+    
+    public void updateMap()
```

Comment:
Should be private since it's only used internally. There is no reason to call it from outside.